src/SDL_properties.c

```diff
@@ -0,0 +1,283 @@
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+#include "SDL_hashtable.h"
+#include "SDL_properties_c.h"
+
+
+typedef struct
+{
+    void *value;
+    void (SDLCALL *cleanup)(void *userdata, void *value);
+    void *userdata;
+} SDL_Property;
+
+typedef struct
+{
+    SDL_HashTable *props;
+    SDL_Mutex *lock;
+} SDL_Properties;
+
+static SDL_HashTable *SDL_properties;
+static SDL_RWLock *SDL_properties_lock;
+static SDL_PropertiesID SDL_last_properties_id;
+
+
+static void SDL_FreeProperty(const void *key, const void *value, void *data)
+{
+    SDL_Property *property = (SDL_Property *)value;
+    if (property->cleanup) {
+        property->cleanup(property->userdata, property->value);
+    }
+    SDL_free((void *)key);
+    SDL_free((void *)value);
+}
+
+static void SDL_FreeProperties(const void *key, const void *value, void *data)
+{
+    SDL_Properties *properties = (SDL_Properties *)value;
+    if (properties) {
+        if (properties->props) {
+            SDL_DestroyHashTable(properties->props);
+            properties->props = NULL;
+        }
+        if (properties->lock) {
+            SDL_DestroyMutex(properties->lock);
+            properties->lock = NULL;
+        }
+        SDL_free(properties);
+    }
+}
+
+int SDL_InitProperties(void)
+{
+    if (!SDL_properties_lock) {
+        SDL_properties_lock = SDL_CreateRWLock();
+        if (!SDL_properties_lock) {
+            return -1;
+        }
+    }
+    if (!SDL_properties) {
+        SDL_properties = SDL_CreateHashTable(NULL, 16, SDL_HashID, SDL_KeyMatchID, SDL_FreeProperties, SDL_FALSE);
+        if (!SDL_properties) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+void SDL_QuitProperties(void)
+{
+    if (SDL_properties) {
+        SDL_DestroyHashTable(SDL_properties);
+        SDL_properties = NULL;
+    }
+    if (SDL_properties_lock) {
+        SDL_DestroyRWLock(SDL_properties_lock);
+        SDL_properties_lock = NULL;
+    }
+}
+
+SDL_PropertiesID SDL_CreateProperties(void)
+{
+    SDL_PropertiesID props;
+    SDL_Properties *properties = NULL;
+    SDL_bool inserted = SDL_FALSE;
+
+    if (!SDL_properties && SDL_InitProperties() < 0) {
+        return 0;
+    }
+
+    properties = SDL_malloc(sizeof(*properties));
+    if (!properties) {
+        goto error;
+    }
+    properties->props = SDL_CreateHashTable(NULL, 4, SDL_HashString, SDL_KeyMatchString, SDL_FreeProperty, SDL_FALSE);
+    if (!properties->props) {
+        goto error;
+    }
+    properties->lock = SDL_CreateMutex();
+    if (!properties->lock) {
+        goto error;
+    }
+
+    if (SDL_InitProperties() < 0) {
+        goto error;
+    }
+
+    if (SDL_LockRWLockForWriting(SDL_properties_lock) == 0) {
+        props = ++SDL_last_properties_id;
```

Comment:
After we create (and destroy) 2<sup>32</sup>-1 properties objects, `SDL_last_properties_id` will overflow and the new properties can have an id of `0`.
This will fail the test in e.g. `SDL_GetSurfaceProperties` and may lead to leaked memory + undefined SDL behavior.

Reply:
Creating properties is really slow: around >4.5ms on my machine. Destroying them is faster: 2µs.
So it would take ~30265 year to hit this overflow.

Is this long delay acceptable?

Reply:
I fixed the wraparound issue, but creating properties shouldn't be that slow. I'll take a look.

Reply:
Can you see why it's so slow on your system? It's as fast as expected here.

In a debug build, running this code:
```
    {
        Uint64 then, now;
        int i;
        then = SDL_GetTicks();
        for (i = 0; i < 100000; ++i) {
            SDL_PropertiesID p = SDL_CreateProperties();
            SDL_DestroyProperties(p);
        }
        now = SDL_GetTicks();
        SDL_Log("100000 creates: %dms\n", (int)(now - then));
    }
```
I get "INFO: 100000 creates: 41m"

Reply:
I'm hitting my head against the wall as I'm writing this.

I was using `SDLTest_TrackAllocations` to try to catch a leak.
I forgot it is ultra slow...

Sorry for taking your time.

Reply:
Oh, hah, okay.